### PR TITLE
Alertmanager: Replicate state using the Ring

### PR DESCRIPTION
**Pull request by gotjosh**
**What this PR does**:

Alertmanager typically uses the memberlist gossip based protocol to
replicate state across replicas. In cortex, we used the same fundamentals
to provide some sort of high availability mode.

Now that we have support for sharding instances across many machines, we
can leverage the ring to find the corresponding instances and send the
updates via gRPC.

This is part of the proposal #3574 

And follows up #3664 , #3671 

Marking it as a draft as I have a few todos left that I need to address but the logic is pretty much what you see on the tin at the moment.

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/alertmanager/alertmanager.go
Comment (pracucci) [inline pkg/alertmanager/alertmanager.go:359]: Question (don't do it, let's just talk about it). Do we need the `SetPosition()`? What if `Alertmanager` was ring-aware and query the ring on demand each time `Position()` is called?

File: pkg/alertmanager/state_replication.go
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:68]: Can be moved below, within the `&state{}` initialisation.

File: pkg/alertmanager/state_replication.go
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:199]: `key`, `logger` and `reg` are unused

File: pkg/alertmanager/state_replication.go
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:52]: Design wise, I'm not a big fan of taking in input `stopc`. I think `state` should self-manage its lifecycler (eg. could be a service).

File: pkg/alertmanager/state_replication.go
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:119]: Why not returning error?

File: pkg/alertmanager/state_replication.go
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:125]: Why not returning error?

File: pkg/alertmanager/state_replication.go
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:40]: What's the impact of returning empty string?

File: pkg/alertmanager/state_replication.go
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:44]: No need to be implemented? If so, comment why.

File: pkg/alertmanager/state_replication.go
Comment (pracucci) [inline pkg/alertmanager/state_replication.go:48]: No need to be implemented? If so, comment why.

File: pkg/alertmanager/alertmanager.go
Comment (pracucci) [inline pkg/alertmanager/alertmanager.go:151]: Why do we need duplicated logic here? I'm wondering if we could refactor `cfg.Peer` to be an interface and we set it either to the alertmanager clustering peer or `am.state` so that we keep the same logic between the two, but the underlying implementation changes.

File: pkg/alertmanager/multitenant.go
Comment (gotjosh) [reply pkg/alertmanager/multitenant.go:429]: Good catch, this was a bad rebase.

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go:236]: Good catch, I managed to get of rid it in favour of a callback

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go:359]: I thought about that but my initial impression is that would be too intense on the ring. We'd basically need to do a lookup for every alert that _tries_ to fire for each tenant across all replicas. 

This seemed like a good enough trade-off to test its term of behaviour and see where it puts us (as this a consistency trade-off)

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:52]: Alertmanager isn't so it seemed like overkill, happy to do it if you think otherwise.

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:119]: Good catch, I was thinking I wouldn't return any merge errors given these messages are supposed to come over UDP (we settle the gossip before sending an alert) but I think it's worth saying something here.

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:125]: Ditto from above.

File: pkg/alertmanager/multitenant.go
Comment (gotjosh) [reply pkg/alertmanager/multitenant.go:901]: I don't think we should error here. [According to the algorithm](https://github.com/prometheus/alertmanager/blob/master/cluster/delegate.go#L150-L168) partial messages are sent over UDP - in this case, we don't care if the message gets lost as we'll settle the gossip before sending an alert.

The trade-off here will be eventual consistency on the API for silences (and maybe alert in the future?) but we're planning on providing a union of results for reads via the distributor.

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:40]: I added a comment on 

```
		// Cortex should not expose cluster information back to its tenants.
		Peer:     &state{},
```

Seems like my attempt to make it more visible was futile, I'll add one to these methods.

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go:151]: I think that's very reasonable, however, we're not done just yet. Given the next step is to implement the:

- Full state replication
- Persistence of state to object storage

I feel like we should wait until those two lands so that we can have an idea of how the final interface looks like.

File: pkg/alertmanager/multitenant.go
Comment (gotjosh) [reply pkg/alertmanager/multitenant.go:902]: My comment about UDP messages and gossip settling apply here. This is a valid use-case that would happen due to a topology change and it wouldn't necessary be an error state.

File: pkg/alertmanager/multitenant.go
Comment (gotjosh) [reply pkg/alertmanager/multitenant.go:750]: There was a test that was screaming at me for this IIRC, but now it seems like it is not...

File: pkg/alertmanager/alertmanager.go
Comment (pracucci) [reply pkg/alertmanager/alertmanager.go]: > would be too intense on the ring

It's not. The ring is an in-memory structure. You can call `.Get()` as many times as you want (we do it thousands time / sec in the distributor).

Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go]: TIL - This makes perfect sense.

File: pkg/alertmanager/multitenant.go
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:892]: The chain of calls leading to this function is a bit complicated. What if we move this function to `state`? You need to pass `selfAddress`, the ring client (not lifecycler) and the clients pool.

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go:204]: Fair, I'll create a [null object](https://en.wikipedia.org/wiki/Null_object_pattern)

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go:189]: In my eyes, this seemed potentially dangerous. Making this call synchronous _might_ significantly slow down the syncing process of the alertmanager depending on how long it takes to settle so I opted for a separate goroutine - seem like one of those details that might be easy to miss can cause undesired side-effects. WDYT?

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go:232]: I'll leave this note, I left it as a reminder that is not really until we call `ApplyConfig` that the alert manager is ready to receive requests but that is currently done synchronously at the moment.

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go]: Done, resolving this conversation.

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go:158]: Adding to this argument is the fact that `Peer`, as defined by upstream, is something that's global - it is always the same for _all_ the alertmanagers thus initialized in the multitenant manager. Whereas in this design `state` is something local to the tenent's alertmanager.

**Comment by jtlisi** [comment]
Fixes #2650

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go]: I took this for a spin - is this better?

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go]: Is this better?

File: pkg/alertmanager/multitenant.go
Comment (stevesg) [inline pkg/alertmanager/multitenant.go:921]:
```suggestion
	return err
```

File: pkg/alertmanager/state_replication.go
Comment (stevesg) [inline pkg/alertmanager/state_replication.go:159]: Did you plan to fix this one before merging, or was it intended for the next commit/commits?

Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:159]: It is intended for the next set of commits - I'm not sure whenever we need a timeout here or what are the implications. There might be a world where background is good enough but didn't want this to go unnoticed.

File: pkg/alertmanager/alertmanager.go
Comment (pracucci) [reply pkg/alertmanager/alertmanager.go:243]: We may remove `WaitReady()` call from this PR to unblock it and address it separately.

Comment (pstibrany) [reply pkg/alertmanager/alertmanager.go:243]: > _Generally speaking I believe it's a bad design "waiting for something" in a constructor function (like this `New()`) but I would like to better understand why we need it._

100%.

(Personally I would like to see this to follow our `Services` model, but I am not sure it makes sense here.)

File: pkg/alertmanager/state_replication.go
Comment (pstibrany) [inline pkg/alertmanager/state_replication.go:120]: Given the comment, does this method need to be part of `State` interface?

File: pkg/alertmanager/state_replication.go
Comment (pstibrany) [inline pkg/alertmanager/state_replication.go:127]: Is this going to be implemented in this PR?

File: pkg/alertmanager/state_replication.go
Comment (pstibrany) [inline pkg/alertmanager/state_replication.go:136]: `WaitReady` needs to be passed a context to solve this.

File: pkg/alertmanager/state_replication.go
Comment (pstibrany) [inline pkg/alertmanager/state_replication.go:124]: Settle ignores context currently. It should not do that, and should report error back if context is finished before settling has finished.

File: pkg/alertmanager/state_replication.go
Comment (pstibrany) [inline pkg/alertmanager/state_replication.go:159]: Hint: If `state` was a `services.Service` (= object with lifecycle in Cortex), it would already have its own context. :)

File: pkg/alertmanager/state_replication.go
Comment (pstibrany) [inline pkg/alertmanager/state_replication.go:58]: Do we need per-key metrics or would per-user (i.e. per-state) be enough?

File: pkg/alertmanager/state_replication.go
Comment (pstibrany) [inline pkg/alertmanager/state_replication.go:190]: This shouldn't be responsibility of "stateChannel", but `replicateStateFunc` (ie.  `MultitenantAlertmanager.replicateStateForUser`), which can track queues per user and implement some fairness, if needed.

File: pkg/alertmanager/state_replication.go
Comment (pstibrany) [inline pkg/alertmanager/state_replication.go:191]: We can move this logging into `MultitenantAlertmanager.replicateStateForUser` and simplify this struct.

File: pkg/alertmanager/alertmanager_metrics.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager_metrics.go:158]: We don't have user as part of these labels because `key` already includes the user.  It's a combination of `prefix + userID`. So by using key we "technically get both". Even though it breaks the nomenclature (of always using `user`).

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:127]: No, this one is not.

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:104]: Is this better?

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:106]: Ditto.

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:120]: It does, I need to update the comment. We actually use the upstream implementation on that gossip-based clustering.

Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:120]: Thanks for pointing this out, it made me realise that a) I needed a test as part of this PR and b) that the above implementation is completely wrong.

File: pkg/alertmanager/multitenant_test.go
Comment (gotjosh) [reply pkg/alertmanager/multitenant_test.go:846]: Done! It was needed.

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:120]: Is this better?

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:124]: I'm leaving the implementation of `WaitReady` and `Settle` to the next PR - is it OK If we skip it for now? These two involve a full state replication and I fear we need the implementation of those to make sense of the big picture.

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:136]: Ditto from my comment below - Moreso, `WaitReady` as defined comes from upstream so we'd need to change that too.

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:58]: per-key includes both user and state so it would always be at most 2 per user.

File: pkg/alertmanager/alertmanager.go
Comment (pstibrany) [inline pkg/alertmanager/alertmanager.go:47]: Nit: I would move this type to the bottom of the file, to keep most important types on the top.

File: pkg/alertmanager/alertmanager_metrics.go
Comment (pstibrany) [reply pkg/alertmanager/alertmanager_metrics.go:158]: We have per-user registries, so we could do per-user output, aggregating over all keys. WDYT?

File: pkg/alertmanager/multitenant.go
Comment (pstibrany) [reply pkg/alertmanager/multitenant.go:880]: If I understand it correctly, moving this to `state` can make it more difficult to implement fairness between different users.

File: pkg/alertmanager/state_replication.go
Comment (pstibrany) [reply pkg/alertmanager/state_replication.go:134]: We don't need to copy `WaitReady` from upstream. We can define our own interface with what makes sense for us, and adapt upstream type to our interface via a wrapper.

Introducing context to upstream's `WaitReady` seems pretty straightforward too.

File: pkg/alertmanager/alertmanager.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager.go:129]: The implementation for `WaitReady` and `Settle` are part of the next PR, if you don't mind I'd like to leave it out of this PR for now.

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:134]: DItto from my comment above, it is certainly possible but I would like to avoid it as part of this PR - The implementation of `WaitReady` and `Settle` for the new replication method are yet to be determined.

File: pkg/alertmanager/alertmanager_metrics.go
Comment (gotjosh) [reply pkg/alertmanager/alertmanager_metrics.go:158]: Is that different from the current approach? An aggregation across all keys is not possible if the keys are per user e.g. `sil:user-3` or `nfl:user-2`.

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:188]: @pracucci had an idea where he wanted to move the implementation of `replicateStateFunc` to the state itself - we'll probably need to have this conversation then. is it OK if it stays here? I think it's worth considering whenever you want to implement it closer to where it is received as opposed to closer before it is replicated.

File: pkg/alertmanager/state_replication.go
Comment (gotjosh) [reply pkg/alertmanager/state_replication.go:157]: Can you help me break this down? 

I have not seen us use services within services (That's different from services with sub-services) and even if we do, at its current state it seems like overkill. It's more code to manage the service, in particular, because the State itself it's satisfied by two different things (our `state` or the upstream `Peer`) which in my eyes brings us little benefits.

**Comment by gotjosh** [comment]
Needs #3903

File: pkg/alertmanager/state_replication.go
Comment (stevesg) [inline pkg/alertmanager/state_replication.go:95]: More of a question than anything, now I've spent more time walking through this code. (Apologies if I'm missing something obvious here...)

Do you think we'll need the flexibility of having the `stateChannel` being passed a function in this way? Could the `stateChannel` just take `s.msgc` and `key`, and the `Broadcast` function be implemented concretely? (Illustrated this in a separate comment)

```suggestion
	return &stateChannel{
		msgc: s.msgc,
		key: key,
	}
```

File: pkg/alertmanager/state_replication.go
Comment (stevesg) [inline pkg/alertmanager/state_replication.go:179]: (see last comment)

```suggestion
// stateChannel allows a state publisher to send messages that will be broadcasted to all other alertmanagers that a tenant
// belongs to.
type stateChannel struct {
	msgc chan *clusterpb.Part
	key  string
}

// Broadcast receives a message to be replicated by the state.
func (c *stateChannel) Broadcast(b []byte) {
	c.msgc <- &clusterpb.Part{Key: c.key, Data: b}
}
```

File: pkg/alertmanager/alertmanager_metrics.go
Comment (pstibrany) [reply pkg/alertmanager/alertmanager_metrics.go:158]: If I understand it correctly, we have per-user registries, which only use "key" label. During `alertmanagerMetrics.Collect`, when then call `SendSumOfCountersWithLabels` with `key` label. We could instead call `SendSumOfCountersPerUser` (eg. `data.SendSumOfCountersPerUser(out, m.partialMerges, "alertmanager_partial_state_merges_total")`), which would return `sum(alertmanager_partial_state_merges_total)` per user-registry, and then add `user` label to the output. I think that would be enough. WDYT?

File: pkg/alertmanager/state_replication.go
Comment (pstibrany) [reply pkg/alertmanager/state_replication.go:156]: I'm only talking about `state` being a service (and not `Peer`), since it runs background process and its lifecycle needs to be managed.

Another possibility is to use custom context initialized in new, and canceled when stopping.

File: pkg/alertmanager/state_replication.go
Comment (pstibrany) [reply pkg/alertmanager/state_replication.go]: ok, let's keep as-is in this PR.

**Review by pracucci — APPROVED** [review]

**Review by pstibrany — APPROVED** [review]
LGTM in current state, with caveat that sharding-support is still work in progress (don't use it yet! :)). Non-sharding code seems to be unaffected though.

**Pull request merged**
